aptos-move/move-examples/dao/nft_dao/sources/nft_dao.move

```diff
@@ -0,0 +1,660 @@
+
+    /// Proposal resource not created
+    const EPRPOSALS_NOT_EXIST_AT_ADDRESS: u64 = 12;
+
+    /// Proposal with specified id doesn't exist
+    const EPRPOSAL_ID_NOT_EXIST: u64 = 13;
+
+    /// DAO already created at this address
+    const EDAO_ADDRESS_USED: u64 = 14;
+
+    /// Already an admin of the DAO
+    const EALREADY_ADMIN_OF_THE_DAO: u64 = 15;
+
+    /// Token already used for creating proposal
+    const ETOKEN_USED_FOR_CREATING_PROPOSAL: u64 = 16;
+
+    /// Constants
+    const PROPOSAL_PENDING: u8 = 0;
+    const PROPOSAL_RESOLVED_PASSED: u8 = 1;
+    const PROPOSAL_RESOLVED_NOT_PASSED: u8 = 2;
+
+    struct DAO has key {
+        /// Name of the DAO
+        name: String,
+        /// The minimum number of total votes (both yes and no) a proposal must have in order to be considered valid. A proposal would still need more yes than no to pass.
+        resolve_threshold: u64,
+        /// The NFT Collection that is used to govern the DAO
+        governance_token: GovernanceToken,
+        /// The voting duration in secs
+        voting_duration: u64,
+        /// Minimum weight for proposal
+        min_required_proposer_voting_power: u64,
+        /// Proposal counter
+        next_proposal_id: u64,
+        /// DAO resource account signer capability
+        dao_signer_capability: SignerCapability,
+        /// DAO admin account
+        admin: address,
+    }
+
+    /// This resource stores all the DAO resource addresses that an admin owns
+    struct OwnedDAO has key {
+        daos: vector<address>,
+    }
+
+    /// The collection should only contains NFTs, where all token name only has 1 maximal and immutable
+    /// The total supply is fixed with the token names.
+    struct GovernanceToken has drop, store {
+        /// The creator address of this NFT collection
+        creator: address,
+        /// The collection name
+        collection: String,
+    }
+
+    /// All proposals
+    struct Proposals has key {
+        proposals: Table<u64, Proposal>,
+    }
+
+    /// Store the general information about a proposal
+    struct Proposal has copy, drop, store {
+        /// Name of the proposal, limiting to 64 chars
+        name: String,
+        /// Description of the proposal, limiting to 512 chars
+        description: String,
+        /// The name of function to be executed
+        function_name: String,
+        /// The function arguments to be executed
+        function_args: PropertyMap,
+        /// The start time of the voting
+        start_time_sec: u64,
+        /// Proposal results, unresolved, passed, not passed
+        resolution: u8,
+        /// final voting count of yes votes
+        final_yes_votes: u64,
+        /// final voting count of no votes
+        final_no_votes: u64,
+    }
+
+    struct ProposalVotingStatistics has key {
+        proposals: Table<u64, VotingStatistics>,
+    }
+
+    struct VotingStatistics has store {
+        /// Total yes votes
+        total_yes: u64,
```

Comment:
Just a note for further thoughts - if we only support NFTs here, wouldn't total_yes be the size of the yes_votes table? Same for total_no

Reply:
I think the weight of NFT could be a fast followup extension. let's keep these two fields in case the weight of NFT will be added.